### PR TITLE
feat: add all content

### PR DESCRIPTION
**Pull request by codyoss**

### CODE REVIEW

**Comment by codyoss** [comment]
@tbpg Would you mind taking a look at this too when you have a chance :)

File: CONTIBUTING.md [hunk 1]
```diff
@@ -1,25 +1,29 @@
 # How to become a contributor and submit your own code
```
Comment (averikitsch) [inline CONTIBUTING.md:1]: fix filename typo

File: README.md [hunk 2]
```diff
@@ -44,65 +43,95 @@ Learn how to use Cloud Code for:
...
+
+   ```bash
+    gcloud auth configure-docker
+    ```
+
+3. Build and push the container using docker:
+
     ```bash
-    
+    docker build . -t gcr.io/$GOOGLE_CLOUD_PROJECT/microservice-template
```
Comment (averikitsch) [inline README.md:75]: I am unable to run this docker image locally

Comment (codyoss) [reply README.md:75]: Hmm, this worked for me via docker. The command I used was `docker run --rm -it -p 8080:8080 -e GOOGLE_CLOUD_PROJECT=$GOOGLE_CLOUD_PROJECT gcr.io/$GOOGLE_CLOUD_PROJECT/microservice-template`. Does that command work, or what error are you getting?

File: CONTIBUTING.md (hunk 1 above)
Comment (codyoss) [reply CONTIBUTING.md]: Whoops, Done.

File: handler.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+func (a *App) Handler(w http.ResponseWriter, r *http.Request) {
+	l, err := jsonlog.NewLogger(fmt.Sprintf("projects/%s", a.projectID))
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	l = l.WithRequest(r)
+	l.Infof("Structured logging example.")
```
[the remark this answers is not in the document]
Comment (codyoss) [reply handler.go]: Sure, we can do that via labels.

File: README.md (hunk 2 above)
Comment (averikitsch) [reply README.md:75]: Ah I wasn't passing in the `$GOOGLE_CLOUD_PROJECT` as an env var to docker run. It doesn't seem to error like it's suppose to. After about maybe 30s I see the error: `2021/06/15 23:55:54 unable to initialize application: unable to detect Project ID from GOOGLE_CLOUD_PROJECT or metadata server: compute: Received 500 `host unreachable`

File: README.md [hunk 4]
```diff
@@ -44,65 +43,95 @@ Learn how to use Cloud Code for:
...
-    The Cloud Build configuration file will build and deploy the containerized service
-    to Cloud Run, run tests managed by NPM, then clean up testing resources. This configuration restricts public
-    access to the test service. Therefore, service accounts need to have the permission to issue Id tokens for request authorization:
+
+    The Cloud Build configuration file will build and deploy the containerized
+    service to Cloud Run, run tests, then clean up testing resources. This
+    configuration restricts public access to the test service. Therefore,
+    service accounts need to have the permission to issue ID tokens for request
+    authorization:
     * Enable Cloud Build and IAM APIs:
```
Comment (tbpg) [inline README.md:120]: Should this be a numbered list?

File: advance.cloudbuild.yaml [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+- id: 'Download dependencies'
+  name: golang
+  args: ['go', 'mod', 'download']
+
+- id: 'Install linting tools'
+  name: golang
+  args: ['go', 'install', 'golang.org/x/tools/cmd/goimports@latest']
+
+- id: 'Lint'
```
Comment (tbpg) [inline advance.cloudbuild.yaml:22]: Is this linting or fixing lint?

File: metadata/metadata.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+// information from the GCP metadata server.
+func ProjectID() (string, error) {
+	return metadata.ProjectID()
+}
+
+// Region returns the region of the Cloud Run service. It fetches this
+// information from the GCP metadata server. The returned value is in the format
+// of: projects/PROJECT_NUMBER/regions/REGION.
+func Region() (string, error) {
+	// TODO: cache this value.
```
Comment (tbpg) [inline metadata/metadata.go:35]: Should we either implement this TODO or remove it?

File: metadata/metadata.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+	if err != nil {
+		return "", nil
+	}
+	return resp, nil
+}
+
+// IDToken returns a TokenSource that yields ID tokens. These tokens can be used
+// to authenticate requests with the Token.SetAuthHeader method.
+func IDToken(ctx context.Context, aud string) (oauth2.TokenSource, error) {
+	idtoken.NewTokenSource(ctx, aud)
```
Comment (tbpg) [inline metadata/metadata.go:46]: Delete.

File: handler.go [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+	"cloud.google.com/go/logging/jsonlog"
+)
+
+func (a *App) Handler(w http.ResponseWriter, r *http.Request) {
+	l, err := jsonlog.NewLogger(fmt.Sprintf("projects/%s", a.projectID))
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	l.WithRequest(r).
```
Comment (tbpg) [inline handler.go:30]: Maybe pull all of this into a method on `*App`? Something like `a.Logger(r)` and under the hood that creates a logger for this request?

File: handler_test.go [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+package main
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+func TestApp_Handler(t *testing.T) {
```
Comment (tbpg) [inline handler_test.go:23]: I think this is usually just `TestHandler`?

File: handler_test.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+	req := httptest.NewRequest("GET", "http://example.com", nil)
+	app := &App{}
+
+	app.Handler(rr, req)
+	if rr.Code != http.StatusOK {
+		t.Errorf("Code = %d, want %d", rr.Code, http.StatusOK)
+	}
+	wantBody := "Hello World!\n"
+	if rr.Body.String() != wantBody {
+		t.Errorf("Body = %q, want %q", rr.Body.String(), wantBody)
```
Comment (tbpg) [inline handler_test.go:34]:
```suggestion
	if got := rr.Body.String(); got != wantBody {
		t.Errorf("Body = %q, want %q", got, wantBody)
```
```suggestion
	if rr.Body.String() != wantBody {
		t.Errorf("Body = %q, want %q", rr.Body.String(), wantBody)
```

File: README.md (hunk 2 above)
Comment (grayside) [reply README.md:75]: One of the patterns we often use is to skip metadata server requests for the current project if that variable is set.

File: advance.cloudbuild.yaml (hunk 5 above)
Comment (codyoss) [reply advance.cloudbuild.yaml:22]: Whoops, yeah the intention was lint. Updated.

File: README.md (hunk 4 above)
Comment (codyoss) [reply README.md:120]: It can be, done.

File: handler.go [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"net/http"
+
+	"cloud.google.com/go/logging/jsonlog"
```
Comment (grayside) [inline handler.go]: I hadn't realized Go had shipped a solution for structured logging. Fantastic!

Is this library GA already?

File: metadata/metadata.go (hunk 6 above)
Comment (grayside) [reply metadata/metadata.go]: My understanding is the metadata library already caches.

File: handler.go (hunk 11 above)
Comment (codyoss) [reply handler.go]: Nope, just cut the preview last week. I will cherry pick to be GA soon though.

File: advance.cloudbuild.yaml [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+  name: golang
+  args: ['goimports', '-w', '.']
+
+- id: 'Run Unit Tests'
+  name: golang
+  args: ['go', 'test', '-v', './...']
+
+# Setup resources for system tests
+- id: 'Build Container Image'
```
[the remark this answers is not in the document]
Comment (codyoss) [reply advance.cloudbuild.yaml:35]: I am not sure if it does, would need to look more into how pack works. But in general deps should be shared across steps thanks to the go-modules volume.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (averikitsch) [reply README.md:75]: Currently the other languages are still using GCR, so let's stick with that for now.

File: advance.cloudbuild.yaml [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  name: 'gcr.io/cloud-builders/gcloud:latest'
+  entrypoint: /bin/bash
+  args:
+    - '-c'
+    - |
+      cat /dev/urandom | LC_CTYPE=C tr -dc 'a-z0-9' | head -c 15 > _short_id # Generate a random 15 character alphanumeric string (lowercase only)
+      gcloud run deploy ${_SERVICE_NAME}-$(cat _short_id) \
+        --image $_GCR_HOSTNAME/$PROJECT_ID/$REPO_NAME/$_SERVICE_NAME:$COMMIT_SHA \
+        --region ${_DEPLOY_REGION} \
+        --platform ${_PLATFORM} \
```
[the remark this answers is not in the document]
Comment (averikitsch) [reply advance.cloudbuild.yaml:60]: Good note, this should be updated in a few places

File: README.md [hunk 14]
```diff
@@ -44,65 +43,94 @@ Learn how to use Cloud Code for:
...
+    docker push gcr.io/$GOOGLE_CLOUD_PROJECT/microservice-template
     ```
-3. Deploy to Cloud Run:
+
+4. Deploy to Cloud Run:
+
     ```bash
-    
+    gcloud run deploy microservice-template \
+      --image gcr.io/$GOOGLE_CLOUD_PROJECT/microservice-template \
```
Comment (averikitsch) [inline README.md:83]:
```suggestion
      --image gcr.io/$GOOGLE_CLOUD_PROJECT/microservice-template
```

File: main.go [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+		log.Fatalf("unable to initialize application: %v", err)
+	}
+	log.Println("starting HTTP server")
+	go func() {
+		if err := app.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			log.Fatalf("Server closed: %v", err)
+		}
+	}()
+
+	// Listen for SIGINT to gracefully shutdown.
```
Comment (averikitsch) [inline main.go:56]: This is nice for local dev, but for Cloud Run  it needs to be SIGTERM

**Review by averikitsch — APPROVED** [review]
lgtm, but I default to the Go experts!

File: main.go (hunk 15 above)
Comment (codyoss) [reply main.go:56]: Looking for both now.

File: .github/CODEOWNERS [hunk 16]
```diff
@@ -0,0 +1 @@
+* @codyoss @averikitsch
```
[the remark this answers is not in the document]
Comment (codyoss) [reply .github/CODEOWNERS:1]: Added cdpe-serverless

File: advance.cloudbuild.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (codyoss) [reply advance.cloudbuild.yaml:60]: updated.

File: advance.cloudbuild.yaml [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+    - push
+    - '$_GCR_HOSTNAME/$PROJECT_ID/$REPO_NAME/$_SERVICE_NAME:$COMMIT_SHA' 
+
+- id: 'Deploy to Cloud Run'
+  name: 'gcr.io/cloud-builders/gcloud:latest'
+  entrypoint: /bin/bash
+  args:
+    - '-c'
+    - |
+      cat /dev/urandom | LC_CTYPE=C tr -dc 'a-z0-9' | head -c 15 > _short_id # Generate a random 15 character alphanumeric string (lowercase only)
```
[the remark this answers is not in the document]
Comment (codyoss) [reply advance.cloudbuild.yaml:56]: Would you like me to update this, if so to what?

File: advance.cloudbuild.yaml [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+      gcloud run deploy ${_SERVICE_NAME}-$(cat _short_id) \
```
Comment (averikitsch) [inline advance.cloudbuild.yaml:57]:
```suggestion
      gcloud run deploy ${_SERVICE_NAME}-$BUILD_ID \
```

File: advance.cloudbuild.yaml (hunk 17 above)
[the remark this answers is not in the document]
Comment (averikitsch) [reply advance.cloudbuild.yaml:56]: See suggestion

**Comment by averikitsch** [comment]
@tbpg Is this Go expert approved?

**Comment by averikitsch** [comment]
@codyoss is this ready to merge?

**Comment by codyoss** [comment]
@averikitsch Yes, I believe so!

**Pull request merged**